scipy/stats/_morestats.py

```diff
@@ -2885,6 +2885,83 @@ def fligner(*samples, center='median', proportiontocut=0.05):
     return FlignerResult(Xsq, pval)
 
 
+@_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
+def _mood_inner_lc(xy, x, diffs, srted, n, m, N) -> float:
+    # obtain the unique values and the counts of each.
+    # "a_i, + b_i, = t_i, for j = 1, ... k", where `k` is the number of unique
+    # classes, and "[t]he number of values associated with the x's and y's in
+    # the jth class will be denoted by a_i, and b_i respectively."
+    # (Mielke, 312)
+    # Prepend `diffs` with a non-zero to indicate that the first element should
+    # be marked as not matching what preceded it.
+    diffs_prep = np.concatenate(([1], diffs))
+    # Unique elements are where the was a difference between elements in the
+    # sorted array
+    uniques = srted[diffs_prep != 0]
+    # The count of each element is the bin size for each set of consecutive
+    # differences where the difference is zero. Replace nonzero differences
+    # with 1 and then use the cumulative sum to count the indices.
+    t = np.bincount(np.cumsum(diffs_prep != 0))[1:]
+    k = len(uniques)
+    js = list(np.arange(1, k + 1))
+
+    # the `b` array mentioned in the paper is not used, outside of the
+    # calculation of `t`, so we do not need to calculate it separately. Here
+    # we calculate `a`. In plain language, `a[i]` is the number of values in
+    # `x` that equal `uniques[i]`.
+    srted_xyx = np.sort(np.concatenate((xy, x)))
+    diffs = np.diff(srted_xyx)
+    diffs_prep = np.concatenate(([1], diffs))
+    xyx_counts = np.bincount(np.cumsum(diffs_prep != 0))[1:]
+    a = xyx_counts - t
+    # "Define .. a_0 = b_0 = t_0 = S_0 = 0" (Mielke 312) so we shift  `a`
+    # and `t` arrays over 1 to allow a first element of 0 to accommodate this
+    # indexing.
+    t = np.concatenate(([0], t))
+    a = np.concatenate(([0], a))
+
+    # S is built from `t`, so it does not need a preceding zero added on.
+    S = np.cumsum(t)
+    # define a copy of `S` with a prepending zero for later use to avoid
+    # indexing
+    S_i_m1 = np.concatenate(([0], S[:-1]))
+
+    # Psi, as defined by the 6th unnumbered equation on page 312 (Mielke).
+    # Note that in the paper there is an error where the denominator `2` is
+    # squared when it should be the entire equation.
+    def psi(indicator):
+        return (indicator - (N + 1)/2)**2
+
+    # define summation range for use in calculation of phi, as seen in sum
+    # in the unnumbered equation on the bottom of page 312 (Mielke).
+    s_lower = [S[jsi - 1] + 1 for jsi in js]
+    s_upper = [S[jsi] + 1 for jsi in js]
+    phi_I = [range(s_lower[idx], s_upper[idx]) for idx in range(k)]
```

Comment:
Can these be vectorized? Something like:

```python
s_lower = S[js - 1] + 1
s_upper = S[js] + 1
psi_I = np.concatenate([np.arange(s_lower[idx], s_upper[idx]) for idx in range(k)], axis=0)
```

Then, I think the code below can be simplified as:

```python
psi_temp = psi(psi_I)
phis = np.sum(psi_temp, axis=1) / t
```

Reply:
I had to remove a bit of the more "fancy" vectorization techniques in order to get this code to port over to pythran correctly. I think that while these operations may be technically supported in Pythran, it can get confused relating to indexing with integers and arrays. 

Once I can get it to port over the Pythran correctly here, I will investigate putting this back in.

Reply:
Update on this: the original vectorization is back.

Reply:
> I had to remove a bit of the more "fancy" vectorization techniques in order to get this code to port over to pythran correctly. I think that while these operations may be technically supported in Pythran, it can get confused relating to indexing with integers and arrays.

I made some benchmarks using Pythran. Looks like with the current implementation, we don't get the benefit of porting over to Pythran unless k is quite large.

![benachmark_mood_2](https://user-images.githubusercontent.com/43181252/171177080-95813451-fd46-4865-b15d-072fbed64fdf.png)

Not sure what's happening there, I added the comment `#pythran export _mood_inner_lc(float[:], float[:], float[:], float[:], int, int, int)` and shifted the function to `_hypotests_pythran`. Did you have this in mind or were you thinking of pythranizing the function in a different way?

<details>

<summary>Code I used to benchmark</summary>

```python
import argparse
from math import fsum
import numpy as np
from scipy.stats import mood
import matplotlib.pyplot as plt

np.random.seed(0)


def setup_func(n_classes, n_samples, use_pythran):
    # number of unique elements
    k = np.random.rand(n_classes)

    # x and y
    x = np.random.choice(k, size=n_samples//2)
    y = np.random.choice(k, size=n_samples//2)

    func = lambda: mood(x, y, use_pythran=use_pythran)

    return func


def benchmark(func):
    import timeit

    # determine number so that 0.2 <= total time < 2.0
    number = 1
    for index in range(0, 10):
        number = 10 ** index
        time_number = timeit.timeit(func, number=number,
                                    globals=globals())
        if time_number >= 0.2:
            break

    time_numbers = timeit.repeat(func, number=number,
                                 repeat=7, globals=globals())
    avg_time = fsum(time_numbers) / len(time_numbers)
    return avg_time


def main(args):
    n_classes, n_samples = args.n_classes, args.n_samples
    use_pythran = args.use_pythran
    func = setup_func(n_classes, n_samples, use_pythran)
    avg_time = benchmark(func)
    print(f"Avg. time taken: {avg_time} seconds.")


def make_plots(args):
    # ks = [10, 100, 1000, 10000]
    # Ns = [(100, 1000), (1000, 10000), (10000,), (100000,)]
    ks = [10, 100, 1000, 10000]
    Ns = [(100000,), (100000,), (100000,), (100000,)]

    fig, ax = plt.subplots()
    xlocs = []
    ksNs = []
    avg_times = []
    avg_times_pythran = []

    i = 0
    for j, k in enumerate(ks):
        for N in Ns[j]:
            print(f"Benchmarking for {k=}, {N=}")
            func = setup_func(k, N, use_pythran=False)
            func_pythran =  setup_func(k, N, use_pythran=True)
            avg_times.append(benchmark(func))
            avg_times_pythran.append(benchmark(func_pythran))
            # avg_times.append(np.random.randint(1,5))
            # avg_times_pythran.append(np.random.randint(1,5))
            xlocs.append(i)
            ksNs.append(f'{k=}, {N=}')
            i += 2
    xlocs = np.asarray(xlocs)
    ax.bar(xlocs, avg_times, width=0.5, color='blue',
           label='python')
    ax.bar(xlocs+0.5, avg_times_pythran, width=0.5,
           color='red', label='pythran')
    ax.set_xticks(xlocs+0.25, ksNs)
    ax.set_title("Mood Benchmarks")
    ax.set_ylabel("Time (s)")
    ax.set_xlabel("Parameters (k=<num classes>, N=<num samples>)")
    ax.legend()
    plt.show()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='mood benchmark.')
    parser.add_argument('--n_classes', '-k', type=int, default=100,
                        help='number of unique elements')
    parser.add_argument('--n_samples', '-N', type=int, default=1000,
                        help='total number of samples in x and y')
    parser.add_argument('--use_pythran', action="store_true", default=False,
                        help="use the pytranized version")
    parser.add_argument('--make_plots', action="store_true", default=False,
                        help="use the pytranized version")

    args = parser.parse_args()

    if args.make_plots:
        make_plots(args)
    else:
        main(args)
```

</details>


<details>

<summary>Diff to add Pythran</summary>

```diff
diff --git a/scipy/stats/_hypotests_pythran.py b/scipy/stats/_hypotests_pythran.py
index dba415e3a..f42e668ac 100644
--- a/scipy/stats/_hypotests_pythran.py
+++ b/scipy/stats/_hypotests_pythran.py
@@ -149,3 +149,79 @@ def _compute_outer_prob_inside_method(m, n, g, h):
             A[maxj - minj:maxj - minj + (lastlen - curlen)] = 1
 
     return A[maxj - minj - 1]
+
+
+#pythran export _mood_inner_lc(float[:], float[:], float[:], float[:], int, int, int)
+def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N):
+    # The below line is required to avoid interger overflow errors in Pythran
+    n, m, N = float(n), float(m), float(N)
+    # Obtain the unique values and their frequencies from the pooled samples.
+    # "a_j, + b_j, = t_j, for j = 1, ... k" where `k` is the number of unique
+    # classes, and "[t]he number of values associated with the x's and y's in
+    # the jth class will be denoted by a_j, and b_j respectively."
+    # (Mielke, 312)
+    # Reuse previously computed sorted array and `diff` arrays to obtain the
+    # unique values and counts. Prepend `diffs` with a non-zero to indicate
+    # that the first element should be marked as not matching what preceded it.
+    diffs_prep = np.concatenate(([1], diffs))
+    # Unique elements are where the was a difference between elements in the
+    # sorted array
+    uniques = sorted_xy[diffs_prep != 0]
+    # The count of each element is the bin size for each set of consecutive
+    # differences where the difference is zero. Replace nonzero differences
+    # with 1 and then use the cumulative sum to count the indices.
+    t = np.bincount(np.cumsum(np.asarray(diffs_prep != 0, dtype=int)))[1:]
+    k = len(uniques)
+    js = np.arange(1, k + 1, dtype=int)
+    # the `b` array mentioned in the paper is not used, outside of the
+    # calculation of `t`, so we do not need to calculate it separately. Here
+    # we calculate `a`. In plain language, `a[j]` is the number of values in
+    # `x` that equal `uniques[j]`.
+    sorted_xyx = np.sort(np.concatenate((xy, x)))
+    diffs = np.diff(sorted_xyx)
+    diffs_prep = np.concatenate(([1], diffs))
+    diff_is_zero = np.asarray(diffs_prep != 0, dtype=int)
+    xyx_counts = np.bincount(np.cumsum(diff_is_zero))[1:]
+    a = xyx_counts - t
+    # "Define .. a_0 = b_0 = t_0 = S_0 = 0" (Mielke 312) so we shift  `a`
+    # and `t` arrays over 1 to allow a first element of 0 to accommodate this
+    # indexing.
+    t = np.concatenate(([0], t))
+    a = np.concatenate(([0], a))
+    # S is built from `t`, so it does not need a preceding zero added on.
+    S = np.cumsum(t)
+    # define a copy of `S` with a prepending zero for later use to avoid
+    # the need for indexing.
+    S_i_m1 = np.concatenate(([0], S[:-1]))
+
+    # Psi, as defined by the 6th unnumbered equation on page 313 (Mielke).
+    # Note that in the paper there is an error where the denominator `2` is
+    # squared when it should be the entire equation.
+    def psi(indicator):
+        return (indicator - (N + 1)/2)**2
+
+    # define summation range for use in calculation of phi, as seen in sum
+    # in the unnumbered equation on the bottom of page 312 (Mielke).
+    s_lower = S[js - 1] + 1
+    s_upper = S[js] + 1
+    phi_J = [np.arange(s_lower[idx], s_upper[idx]) for idx in range(k)]
+
+    # for every range in the above array, determine the sum of psi(I) for
+    # every element in the range. Divide all the sums by `t`. Following the
+    # last unnumbered equation on page 312.
+    phis = [np.sum(psi(I_j)) for I_j in phi_J] / t[js]
+
+    # `T` is equal to a[j] * phi[j], per the first unnumbered equation on
+    # page 312. `phis` is already in the order based on `js`, so we index
+    # into `a` with `js` as well.
+    T = sum(phis * a[js])
+
+    # The approximate statistic
+    E_0_T = n * (N * N - 1) / 12
+
+    varM = (m * n * (N + 1.0) * (N ** 2 - 4) / 180 -
+            m * n / (180 * N * (N - 1)) * np.sum(
+                t * (t**2 - 1) * (t**2 - 4 + (15 * (N - S - S_i_m1) ** 2))
+            ))
+
+    return ((T - E_0_T) / np.sqrt(varM),)
diff --git a/scipy/stats/_morestats.py b/scipy/stats/_morestats.py
index 6b0b48d84..6f50da1a3 100644
--- a/scipy/stats/_morestats.py
+++ b/scipy/stats/_morestats.py
@@ -22,6 +22,7 @@ from . import distributions
 from ._distn_infrastructure import rv_generic
 from ._hypotests import _get_wilcoxon_distr
 from ._axis_nan_policy import _axis_nan_policy_factory
+from ._hypotests_pythran import _mood_inner_lc
 
 
 __all__ = ['mvsdist',
@@ -2886,7 +2887,7 @@ def fligner(*samples, center='median', proportiontocut=0.05):
 
 
 @_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
-def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
+def _mood_inner_lc_vanilla(xy, x, diffs, sorted_xy, n, m, N) -> float:
     # Obtain the unique values and their frequencies from the pooled samples.
     # "a_j, + b_j, = t_j, for j = 1, ... k" where `k` is the number of unique
     # classes, and "[t]he number of values associated with the x's and y's in
@@ -2959,7 +2960,12 @@ def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
     return ((T - E_0_T) / np.sqrt(varM),)
 
 
-def mood(x, y, axis=0, alternative="two-sided"):
+@_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
+def _mood_inner_lc_pythranized(xy, x, diffs, sorted_xy, n, m, N):
+    return _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N)
+
+
+def mood(x, y, axis=0, alternative="two-sided", use_pythran=False):
     """Perform Mood's test for equal scale parameters.
 
     Mood's two-sample test for scale parameters is a non-parametric
@@ -3072,8 +3078,11 @@ def mood(x, y, axis=0, alternative="two-sided"):
     sorted_xy = np.sort(xy, axis=axis)
     diffs = np.diff(sorted_xy, axis=axis)
     if 0 in diffs:
-        z = np.asarray(_mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N,
-                                      axis=axis))
+        _mood_inner_lc_func = _mood_inner_lc_vanilla
+        if use_pythran:
+            _mood_inner_lc_func = _mood_inner_lc_pythranized
+        z = np.asarray(_mood_inner_lc_func(xy, x, diffs, sorted_xy, n, m, N,
+                                           axis=axis))
     else:
         if axis != 0:
             xy = np.moveaxis(xy, axis, 0)
(END)
+            ))
+
+    return ((T - E_0_T) / np.sqrt(varM),)
diff --git a/scipy/stats/_morestats.py b/scipy/stats/_morestats.py
index 6b0b48d84..6f50da1a3 100644
--- a/scipy/stats/_morestats.py
+++ b/scipy/stats/_morestats.py
@@ -22,6 +22,7 @@ from . import distributions
 from ._distn_infrastructure import rv_generic
 from ._hypotests import _get_wilcoxon_distr
 from ._axis_nan_policy import _axis_nan_policy_factory
+from ._hypotests_pythran import _mood_inner_lc
 
 
 __all__ = ['mvsdist',
@@ -2886,7 +2887,7 @@ def fligner(*samples, center='median', proportiontocut=0.05):
:
+            ))
+
+    return ((T - E_0_T) / np.sqrt(varM),)
diff --git a/scipy/stats/_morestats.py b/scipy/stats/_morestats.py
index 6b0b48d84..6f50da1a3 100644
--- a/scipy/stats/_morestats.py
+++ b/scipy/stats/_morestats.py
@@ -22,6 +22,7 @@ from . import distributions
 from ._distn_infrastructure import rv_generic
 from ._hypotests import _get_wilcoxon_distr
 from ._axis_nan_policy import _axis_nan_policy_factory
+from ._hypotests_pythran import _mood_inner_lc
 
 
 __all__ = ['mvsdist',
@@ -2886,7 +2887,7 @@ def fligner(*samples, center='median', proportiontocut=0.05):
 
 
:
+            ))
+
+    return ((T - E_0_T) / np.sqrt(varM),)
diff --git a/scipy/stats/_morestats.py b/scipy/stats/_morestats.py
index 6b0b48d84..6f50da1a3 100644
--- a/scipy/stats/_morestats.py
+++ b/scipy/stats/_morestats.py
@@ -22,6 +22,7 @@ from . import distributions
 from ._distn_infrastructure import rv_generic
 from ._hypotests import _get_wilcoxon_distr
 from ._axis_nan_policy import _axis_nan_policy_factory
+from ._hypotests_pythran import _mood_inner_lc
 
 
 __all__ = ['mvsdist',
@@ -2886,7 +2887,7 @@ def fligner(*samples, center='median', proportiontocut=0.05):
 
 
 @_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
-def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
+def _mood_inner_lc_vanilla(xy, x, diffs, sorted_xy, n, m, N) -> float:
     # Obtain the unique values and their frequencies from the pooled samples.
     # "a_j, + b_j, = t_j, for j = 1, ... k" where `k` is the number of unique
     # classes, and "[t]he number of values associated with the x's and y's in
@@ -2959,7 +2960,12 @@ def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
     return ((T - E_0_T) / np.sqrt(varM),)
 
 
-def mood(x, y, axis=0, alternative="two-sided"):
+@_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
+def _mood_inner_lc_pythranized(xy, x, diffs, sorted_xy, n, m, N):
+    return _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N)
+
+
+def mood(x, y, axis=0, alternative="two-sided", use_pythran=False):
     """Perform Mood's test for equal scale parameters.
 
     Mood's two-sample test for scale parameters is a non-parametric
@@ -3072,8 +3078,11 @@ def mood(x, y, axis=0, alternative="two-sided"):
     sorted_xy = np.sort(xy, axis=axis)
     diffs = np.diff(sorted_xy, axis=axis)
     if 0 in diffs:
-        z = np.asarray(_mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N,
-                                      axis=axis))
+        _mood_inner_lc_func = _mood_inner_lc_vanilla
+        if use_pythran:
+            _mood_inner_lc_func = _mood_inner_lc_pythranized
+        z = np.asarray(_mood_inner_lc_func(xy, x, diffs, sorted_xy, n, m, N,
+                                           axis=axis))
     else:
         if axis != 0:
             xy = np.moveaxis(xy, axis, 0)
~
~
~
(END)
+            ))
+
+    return ((T - E_0_T) / np.sqrt(varM),)
diff --git a/scipy/stats/_morestats.py b/scipy/stats/_morestats.py
index 6b0b48d84..6f50da1a3 100644
--- a/scipy/stats/_morestats.py
+++ b/scipy/stats/_morestats.py
@@ -22,6 +22,7 @@ from . import distributions
 from ._distn_infrastructure import rv_generic
 from ._hypotests import _get_wilcoxon_distr
 from ._axis_nan_policy import _axis_nan_policy_factory
+from ._hypotests_pythran import _mood_inner_lc
 
 
 __all__ = ['mvsdist',
@@ -2886,7 +2887,7 @@ def fligner(*samples, center='median', proportiontocut=0.05):
 
 
 @_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
-def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
+def _mood_inner_lc_vanilla(xy, x, diffs, sorted_xy, n, m, N) -> float:
     # Obtain the unique values and their frequencies from the pooled samples.
     # "a_j, + b_j, = t_j, for j = 1, ... k" where `k` is the number of unique
     # classes, and "[t]he number of values associated with the x's and y's in
@@ -2959,7 +2960,12 @@ def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
     return ((T - E_0_T) / np.sqrt(varM),)
 
 
-def mood(x, y, axis=0, alternative="two-sided"):
+@_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
+def _mood_inner_lc_pythranized(xy, x, diffs, sorted_xy, n, m, N):
+    return _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N)
+
+
+def mood(x, y, axis=0, alternative="two-sided", use_pythran=False):
     """Perform Mood's test for equal scale parameters.
 
     Mood's two-sample test for scale parameters is a non-parametric
@@ -3072,8 +3078,11 @@ def mood(x, y, axis=0, alternative="two-sided"):
     sorted_xy = np.sort(xy, axis=axis)
     diffs = np.diff(sorted_xy, axis=axis)
     if 0 in diffs:
-        z = np.asarray(_mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N,
-                                      axis=axis))
+        _mood_inner_lc_func = _mood_inner_lc_vanilla
+        if use_pythran:
+            _mood_inner_lc_func = _mood_inner_lc_pythranized
+        z = np.asarray(_mood_inner_lc_func(xy, x, diffs, sorted_xy, n, m, N,
+                                           axis=axis))
     else:
         if axis != 0:
             xy = np.moveaxis(xy, axis, 0)
```

</details>

Reply:
Thanks for taking such a close look, this wasn't the result I had previously experienced. While I take a look as well, can you confirm that Pythran is working / enabled on your development environment? I know that @mdhaber let me know that he had trouble getting Pythran to work on windows, and I initially forgot to set `USE_PYTHRAN=1` when I was setting up mine.

Reply:
> Thanks for taking such a close look, this wasn't the result I had previously experienced. While I take a look as well, can you confirm that Pythran is working / enabled on your development environment? I know that @mdhaber let me know that he had trouble getting Pythran to work on windows, and I initially forgot to set `USE_PYTHRAN=1` when I was setting up mine.

Yes, Pythran is enabled in my environment.

Reply:
Thanks, 

I have run your benchmarking file as well after making the same changes to my repo. Weirdly, I get almost opposite results, that performance is only better with pythran for small `k` values. So I think I need to keep investigating. Maybe it is system dependent somehow.


![image](https://user-images.githubusercontent.com/44255917/171228471-5af62861-e0f7-413f-aeda-2a7553719f7b.png)



Previously I had only tested with smaller cases with relatively few ties, like this one: 
```
In [10]: x = [111, 107, 100, 99, 102, 106, 109, 108, 104, 99,
    ...:      101, 96, 97, 102, 107, 113, 116, 113, 110, 98]
    ...: y = [107, 108, 106, 98, 105, 103, 110, 105, 104, 100,
    ...:      96, 108, 103, 104, 114, 114, 113, 108, 106, 99]

In [11]: %timeit mood(x, y, use_pythran=False)
1.16 ms ± 88.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [12]: %timeit mood(x, y, use_pythran=True)
594 µs ± 58.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Reply:
> Maybe it is system dependent somehow.

Yep, that's my first thought too. But I think Pythran is not able to optimize bincount and some other NumPy functions. I tried by factoring out the two list comprehensions which were the bottleneck of `_mood_inner_lc` and it behaves exactly as expected:

![benachmark_mood_3](https://user-images.githubusercontent.com/43181252/171235888-9a368538-8fde-4c9f-9eca-dd7365da2039.png)

<details>
<summary>New diff</summary>

```diff
diff --git a/scipy/stats/_hypotests_pythran.py b/scipy/stats/_hypotests_pythran.py
index dba415e3a..be1e1bbc0 100644
--- a/scipy/stats/_hypotests_pythran.py
+++ b/scipy/stats/_hypotests_pythran.py
@@ -149,3 +149,96 @@ def _compute_outer_prob_inside_method(m, n, g, h):
             A[maxj - minj:maxj - minj + (lastlen - curlen)] = 1
 
     return A[maxj - minj - 1]
+
+
+#pythran export _mood_inner_lc(float[:], float[:], float[:], float[:], int, int, int)
+def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N):
+    # The below line is required to avoid interger overflow errors in Pythran
+    n, m, N = float(n), float(m), float(N)
+    # Obtain the unique values and their frequencies from the pooled samples.
+    # "a_j, + b_j, = t_j, for j = 1, ... k" where `k` is the number of unique
+    # classes, and "[t]he number of values associated with the x's and y's in
+    # the jth class will be denoted by a_j, and b_j respectively."
+    # (Mielke, 312)
+    # Reuse previously computed sorted array and `diff` arrays to obtain the
+    # unique values and counts. Prepend `diffs` with a non-zero to indicate
+    # that the first element should be marked as not matching what preceded it.
+    diffs_prep = np.concatenate(([1], diffs))
+    # Unique elements are where the was a difference between elements in the
+    # sorted array
+    uniques = sorted_xy[diffs_prep != 0]
+    # The count of each element is the bin size for each set of consecutive
+    # differences where the difference is zero. Replace nonzero differences
+    # with 1 and then use the cumulative sum to count the indices.
+    t = np.bincount(np.cumsum(np.asarray(diffs_prep != 0, dtype=int)))[1:]
+    k = len(uniques)
+    js = np.arange(1, k + 1, dtype=int)
+    # the `b` array mentioned in the paper is not used, outside of the
+    # calculation of `t`, so we do not need to calculate it separately. Here
+    # we calculate `a`. In plain language, `a[j]` is the number of values in
+    # `x` that equal `uniques[j]`.
+    sorted_xyx = np.sort(np.concatenate((xy, x)))
+    diffs = np.diff(sorted_xyx)
+    diffs_prep = np.concatenate(([1], diffs))
+    diff_is_zero = np.asarray(diffs_prep != 0, dtype=int)
+    xyx_counts = np.bincount(np.cumsum(diff_is_zero))[1:]
+    a = xyx_counts - t
+    # "Define .. a_0 = b_0 = t_0 = S_0 = 0" (Mielke 312) so we shift  `a`
+    # and `t` arrays over 1 to allow a first element of 0 to accommodate this
+    # indexing.
+    t = np.concatenate(([0], t))
+    a = np.concatenate(([0], a))
+    # S is built from `t`, so it does not need a preceding zero added on.
+    S = np.cumsum(t)
+    # define a copy of `S` with a prepending zero for later use to avoid
+    # the need for indexing.
+    S_i_m1 = np.concatenate(([0], S[:-1]))
+
+    # Psi, as defined by the 6th unnumbered equation on page 313 (Mielke).
+    # Note that in the paper there is an error where the denominator `2` is
+    # squared when it should be the entire equation.
+    def psi(indicator):
+        return (indicator - (N + 1)/2)**2
+
+    # define summation range for use in calculation of phi, as seen in sum
+    # in the unnumbered equation on the bottom of page 312 (Mielke).
+    s_lower = S[js - 1] + 1
+    s_upper = S[js] + 1
+    phi_J = [np.arange(s_lower[idx], s_upper[idx]) for idx in range(k)]
+
+    # for every range in the above array, determine the sum of psi(I) for
+    # every element in the range. Divide all the sums by `t`. Following the
+    # last unnumbered equation on page 312.
+    phis = [np.sum(psi(I_j)) for I_j in phi_J] / t[js]
+
+    # `T` is equal to a[j] * phi[j], per the first unnumbered equation on
+    # page 312. `phis` is already in the order based on `js`, so we index
+    # into `a` with `js` as well.
+    T = sum(phis * a[js])
+
+    # The approximate statistic
+    E_0_T = n * (N * N - 1) / 12
+
+    varM = (m * n * (N + 1.0) * (N ** 2 - 4) / 180 -
+            m * n / (180 * N * (N - 1)) * np.sum(
+                t * (t**2 - 1) * (t**2 - 4 + (15 * (N - S - S_i_m1) ** 2))
+            ))
+
+    return ((T - E_0_T) / np.sqrt(varM),)
+
+
+#pythran export _mood_phi(int[:], int[:], int[:], int[:], int, int)
+def _mood_phi(s_lower, s_upper, t, js, k, N):
+    # Psi, as defined by the 6th unnumbered equation on page 313 (Mielke).
+    # Note that in the paper there is an error where the denominator `2` is
+    # squared when it should be the entire equation.
+    def psi(I):
+        return (I - (N + 1)/2)**2
+
+    phi_J = [np.arange(s_lower[idx], s_upper[idx]) for idx in range(k)]
+
+    # for every range in the above array, determine the sum of psi(I) for
+    # every element in the range. Divide all the sums by `t`. Following the
+    # last unnumbered equation on page 312.
+    phis = [np.sum(psi(I_j)) for I_j in phi_J] / t[js]
+    return phis
diff --git a/scipy/stats/_morestats.py b/scipy/stats/_morestats.py
index 6b0b48d84..d899e93f4 100644
--- a/scipy/stats/_morestats.py
+++ b/scipy/stats/_morestats.py
@@ -22,6 +22,7 @@ from . import distributions
 from ._distn_infrastructure import rv_generic
 from ._hypotests import _get_wilcoxon_distr
 from ._axis_nan_policy import _axis_nan_policy_factory
+from ._hypotests_pythran import _mood_inner_lc, _mood_phi
 
 
 __all__ = ['mvsdist',
@@ -2886,7 +2887,7 @@ def fligner(*samples, center='median', proportiontocut=0.05):
 
 
 @_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
-def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
+def _mood_inner_lc_vanilla(xy, x, diffs, sorted_xy, n, m, N) -> float:
     # Obtain the unique values and their frequencies from the pooled samples.
     # "a_j, + b_j, = t_j, for j = 1, ... k" where `k` is the number of unique
     # classes, and "[t]he number of values associated with the x's and y's in
@@ -2926,27 +2927,17 @@ def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
     # the need for indexing.
     S_i_m1 = np.concatenate(([0], S[:-1]))
 
-    # Psi, as defined by the 6th unnumbered equation on page 313 (Mielke).
-    # Note that in the paper there is an error where the denominator `2` is
-    # squared when it should be the entire equation.
-    def psi(indicator):
-        return (indicator - (N + 1)/2)**2
-
     # define summation range for use in calculation of phi, as seen in sum
     # in the unnumbered equation on the bottom of page 312 (Mielke).
     s_lower = S[js - 1] + 1
     s_upper = S[js] + 1
-    phi_J = [np.arange(s_lower[idx], s_upper[idx]) for idx in range(k)]
 
-    # for every range in the above array, determine the sum of psi(I) for
-    # every element in the range. Divide all the sums by `t`. Following the
-    # last unnumbered equation on page 312.
-    phis = [np.sum(psi(I_j)) for I_j in phi_J] / t[js]
+    phis = _mood_phi(s_lower, s_upper, t, js, k, N)
 
     # `T` is equal to a[j] * phi[j], per the first unnumbered equation on
     # page 312. `phis` is already in the order based on `js`, so we index
     # into `a` with `js` as well.
-    T = sum(phis * a[js])
+    T = np.sum(phis * a[js])
 
     # The approximate statistic
     E_0_T = n * (N * N - 1) / 12
@@ -2959,7 +2950,12 @@ def _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N) -> float:
     return ((T - E_0_T) / np.sqrt(varM),)
 
 
-def mood(x, y, axis=0, alternative="two-sided"):
+@_axis_nan_policy_factory(lambda x1: (x1,), n_samples=4, n_outputs=1)
+def _mood_inner_lc_pythranized(xy, x, diffs, sorted_xy, n, m, N):
+    return _mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N)
+
+
+def mood(x, y, axis=0, alternative="two-sided", use_pythran=False):
     """Perform Mood's test for equal scale parameters.
 
     Mood's two-sample test for scale parameters is a non-parametric
@@ -3072,8 +3068,11 @@ def mood(x, y, axis=0, alternative="two-sided"):
     sorted_xy = np.sort(xy, axis=axis)
     diffs = np.diff(sorted_xy, axis=axis)
     if 0 in diffs:
-        z = np.asarray(_mood_inner_lc(xy, x, diffs, sorted_xy, n, m, N,
-                                      axis=axis))
+        _mood_inner_lc_func = _mood_inner_lc_vanilla
+        if use_pythran:
+            _mood_inner_lc_func = _mood_inner_lc_pythranized
+        z = np.asarray(_mood_inner_lc_func(xy, x, diffs, sorted_xy, n, m, N,
+                                           axis=axis))
     else:
         if axis != 0:
             xy = np.moveaxis(xy, axis, 0)
```


</details>

Reply:
> > Maybe it is system dependent somehow.


> 
> Yep, that's my first thought too. But I think Pythran is not able to optimize bincount and some other NumPy functions. I tried by factoring out the two list comprehensions which were the bottleneck of _mood_inner_lc and it behaves exactly as expected:

I see in the [Pythran docs's "Supported Modules and Functions" NumPy's bincount](https://pythran.readthedocs.io/en/latest/SUPPORT.html#numpy) is included , however I don't know if that just means that they permit its usage. 

just to add to this confusion, I can apply your patch and these are the results I get: (opposite again!)
![image](https://user-images.githubusercontent.com/44255917/171246064-342df3e2-d3a9-44b9-99ca-17fdbb714def.png)



Even more confusing, I noticed that every time I run it, I get slightly different results:
![image](https://user-images.githubusercontent.com/44255917/171251085-7a997843-9d50-4012-a97b-e45ec207bf5d.png)
![image](https://user-images.githubusercontent.com/44255917/171251172-8adc9d3d-6d3e-43ee-98ff-6e7dbb6eeee1.png)
![image](https://user-images.githubusercontent.com/44255917/171251254-9c5519f5-3229-4d99-adce-ac8171fc0a2d.png)


Maybe I will make some ASV benchmarks to see if this happens there too.

Reply:
Here's what I get:
<img width="338" alt="image" src="https://user-images.githubusercontent.com/6570539/171294018-7f293f18-af1f-46a3-b7a9-3f6b5e17739e.png">

Reply:
So strange. How about we table the future performance enhancement for a follow up PR so that we can merge in the fix for this defect a little quicker? Then later I can do a much more in depth investigation of what really needs to be in Pythran.

Reply:
I made a silly mistake when writing the code: timeit.repeat returns the time taken to run the code `number` number of times. So, we had to divide the result by `number` to get the correct results. Sorry for the fuss! Pythan is faster:

![benachmark_mood_4](https://user-images.githubusercontent.com/43181252/171353723-39d76e96-801f-44f6-9e8f-0490061fabd4.png)

Feel free to Pythranize in the PR or a follow-up.

Reply:
Resolving this since vectorization looks good.